### PR TITLE
[Optimization][KG] Several optimizations on DGL-KG

### PR DESCRIPTION
**Pull request by classicsong**
## Brief
Several optimizations on DGL-KG:
1. Sorted positive edges for sampling which can reduce random memory access during positive sampling
2. Asynchronous node embedding update
3. Balanced Relation Partition that gives balanced number of edges in each partition. When there is no cross partition relation, relation embedding can be pin into GPU memory
4. Tunable neg_sample_size instead of fixed neg_sample_size

Related Issue: #1229 

### Evaluation (mix_cpu_gpu)
|  TransE_l1 (dim=400)  |  MR   |  MRR  | HITS@1 | HITS@3 | HITS@10 | Speed(sec) |
|-----------|-------|-------|--------|--------|---------|---------|
|  1 GPU | 44.67 | 0.676 | 0.551  | 0.776 | 0.861 | 471.1 |
|  8 GPU | 44.57 | 0.667 | 0.538  | 0.770  | 0.859 | 95.16 |

|  Complex (dim=400)  |  MR   |  MRR  | HITS@1 | HITS@3 | HITS@10 | Speed(sec) |
|-----------|-------|-------|--------|--------|---------|---------|
|  1 GPU | 64.37 | 0.771 | 0.695  | 0.831 | 0.888 | 379.4 |
|  8 GPU | 68.33 | 0.759 | 0.679  | 0.822  | 0.885 | 73.85 |

|  DistMult (dim=400)  |  MR   |  MRR  | HITS@1 | HITS@3 | HITS@10 | Speed(sec) |
|-----------|-------|-------|--------|--------|---------|---------|
|  1 GPU | 61.03 | 0.727 | 0.627  | 0.806 | 0.883 | 433.8 |
|  8 GPU | 59.88 | 0.712 | 0.607  | 0.795  | 0.880 | 82.98 |

|  RotatE (dim=200)  |  MR   |  MRR  | HITS@1 | HITS@3 | HITS@10 | Speed(sec) |
|-----------|-------|-------|--------|--------|---------|---------|
|  1 GPU | 44.80 | 0.716 | 0.618  | 0.790 | 0.867 | 1749.2 |
|  8 GPU | 45.09 | 0.704 | 0.601  | 0.783  | 0.863 | 252.5 |

|  TransR (dim=400)  |  MR   |  MRR  | HITS@1 | HITS@3 | HITS@10 | Speed(sec) |
|-----------|-------|-------|--------|--------|---------|---------|
|  1 GPU | 60.78 | 0.674 | 0.589  | 0.733 | 0.814 | 892.90 |
|  8 GPU | 65.29 | 0.672 | 0.587  | 0.730  | 0.810 | 134.59 |

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: apps/kg/train_pytorch.py [hunk 1]
```diff
@@ -53,10 +29,7 @@ def load_model_from_checkpoint(logger, args, n_entities, n_relations, ckpt_path)
     model.load_emb(ckpt_path, args.dataset)
     return model
 
-@thread_wrapped_func
-def train(args, model, train_sampler, rank=0, rel_parts=None, valid_samplers=None):
-    if args.num_proc > 1:
-        th.set_num_threads(4)
```
Comment (zheng-da) [inline apps/kg/train_pytorch.py]: where to set the number of threads?

File: src/graph/sampler.cc [hunk 2]
```diff
@@ -1227,13 +1227,7 @@ NegSubgraph EdgeSamplerObject::genNegEdgeSubgraph(const Subgraph &pos_subg,
   neg_subg.graph = GraphPtr(new ImmutableGraph(neg_coo));
   neg_subg.induced_vertices = induced_neg_vid;
   neg_subg.induced_edges = induced_neg_eid;
-  // If we didn't sample all nodes to form negative edges, some of the nodes
-  // in the vector might be redundant.
-  if (neg_sample_size < num_tot_nodes) {
-    std::sort(neg_vids.begin(), neg_vids.end());
-    auto it = std::unique(neg_vids.begin(), neg_vids.end());
-    neg_vids.resize(it - neg_vids.begin());
-  }
```
Comment (zheng-da) [inline src/graph/sampler.cc]: should we simply remove it? the subgraph will have duplicated nodes if we just remove it.

File: apps/kg/dataloader/sampler.py [hunk 3]
```diff
@@ -160,15 +216,13 @@ def create_neg_subgraph(pos_g, neg_g, chunk_size, is_chunked, neg_head, num_node
         chunk_size = pos_g.number_of_edges()
     elif is_chunked:
         if pos_g.number_of_edges() < chunk_size:
-            num_chunks = 1
-            chunk_size = pos_g.number_of_edges()
+            return None
         else:
             # This is probably the last batch. Let's ignore it.
             if pos_g.number_of_edges() % chunk_size > 0:
```
[the remark this answers is not in the document]
Comment (classicsong) [reply apps/kg/dataloader/sampler.py]: OK

File: apps/kg/eval.py [hunk 4]
```diff
@@ -100,7 +100,8 @@ def main(args):
     args.train = False
     args.valid = False
     args.test = True
-    args.rel_part = False
+    args.strict_rel_part = False
```
[the remark this answers is not in the document]
Comment (classicsong) [reply apps/kg/eval.py]: Currently, strict_rel_part is a best effort action by KGE atomically. If there is no relation across multipartions, we will use strict_rel_part and pin relation in GPU. (also a default case for 1GPU training)

File: apps/kg/dataloader/sampler.py [hunk 5]
```diff
@@ -270,29 +326,8 @@ def get_edges(self, eval_type):
...
-        orig_src = np.array([t[0] for t in data])
-        orig_etype = np.array([t[1] for t in data])
-        orig_dst = np.array([t[2] for t in data])
-        np.testing.assert_equal(F.asnumpy(src_id), orig_src)
-        np.testing.assert_equal(F.asnumpy(dst_id), orig_dst)
-        np.testing.assert_equal(F.asnumpy(etype), orig_etype)
-
     def create_sampler(self, eval_type, batch_size, neg_sample_size, neg_chunk_size,
-                       filter_false_neg, mode='head', num_workers=5, rank=0, ranks=1):
+                       filter_false_neg, mode='head', num_workers=1, rank=0, ranks=1):
```
[the remark this answers is not in the document]
Comment (classicsong) [reply apps/kg/dataloader/sampler.py]: I only think 5 is weird and there is no best config for all cases. Let's set it to 1 and let user give it through args

File: apps/kg/train_pytorch.py (hunk 1 above)
Comment (classicsong) [reply apps/kg/train_pytorch.py]: There is a func train_mp

File: src/graph/sampler.cc (hunk 2 above)
Comment (classicsong) [reply src/graph/sampler.cc]: It seems OK for subgraph to have duplicated nodes. If we remove the duplicated nodes, we should sample more nodes to make the num_of_edges aligned. 
I think we need another PR to cleanup our neg-sampling code.

File: apps/kg/models/pytorch/tensor_models.py [hunk 6]
```diff
@@ -42,6 +93,8 @@ def __init__(self, args, num, dim, device):
         self.emb = th.empty(num, dim, dtype=th.float32, device=device)
         self.state_sum = self.emb.new().resize_(self.emb.size(0)).zero_()
         self.state_step = 0
+        self.async_q = None
+        self.async_p = None
```
[the remark this answers is not in the document]
Comment (classicsong) [reply apps/kg/models/pytorch/tensor_models.py]: Let me add some some doc strings

File: apps/kg/dataloader/sampler.py [hunk 7]
```diff
@@ -10,7 +10,15 @@
 
 # This partitions a list of edges based on relations to make sure
 # each partition has roughly the same number of edges and relations.
-def RelationPartition(edges, n):
+# Algo:
+#     For r in relations:
+#       Find partition with maximum empty slot
```
[the remark this answers is not in the document]
Comment (classicsong) [reply apps/kg/dataloader/sampler.py]: yes, let me change the description

File: apps/kg/eval.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply apps/kg/eval.py]: strict rel part isn't always the best option. if a partition is too small, strict rel partition cause imbalance. in this case, we should automatically switch to balanced partition.

File: apps/kg/dataloader/sampler.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply apps/kg/dataloader/sampler.py]: I think a large number is always preferred. There isn't too much drawback to use a large number.

[the remark this answers is not in the document]
Comment (zheng-da) [reply apps/kg/dataloader/sampler.py]: if we set 1 by default, who should be responsible for setting the number correctly? The users of DGL-ke package?

File: apps/kg/eval.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (classicsong) [reply apps/kg/eval.py]: Users can use random partition. (without specify --rel_part in args)
If user choose to use --rel_part, the KG package will decide whether it is strict_rel_part or not.

File: apps/kg/dataloader/sampler.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (classicsong) [reply apps/kg/dataloader/sampler.py]: I c. We can change it to 4 or 8.

File: apps/kg/eval.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply apps/kg/eval.py]: I see. you are already doing it

File: apps/kg/models/pytorch/tensor_models.py [hunk 8]
```diff
@@ -76,33 +191,70 @@ def update(self, gpu_id=-1):
...
+                    self.state_sum.index_add_(0, grad_indices, grad_sum)
+                    std = self.state_sum[grad_indices]  # _sparse_mask
+                    if gpu_id >= 0:
+                        std = std.cuda(gpu_id)
+                    std_values = std.sqrt_().add_(1e-10).unsqueeze(1)
+                    tmp = (-clr * grad_values / std_values)
+                    if tmp.device != device:
+                        tmp = tmp.to(device)
+                    # TODO(zhengda) the overhead is here.
+                    self.emb.index_add_(0, grad_indices, tmp)
```
Comment (zheng-da) [inline apps/kg/models/pytorch/tensor_models.py]: looks like this piece of code is replicated?

File: apps/kg/train_pytorch.py [hunk 9]
```diff
@@ -90,52 +68,78 @@ def train(args, model, train_sampler, rank=0, rel_parts=None, valid_samplers=Non
...
             for k, v in metrics.items():
-                print('{} average {} at [{}/{}]: {}'.format(mode, k, args.step, args.max_step, v))
+                print('[{}]{} average {} at [{}/{}]: {}'.format(rank, mode, k, args.step, args.max_step, v))
     test_samplers[0] = test_samplers[0].reset()
     test_samplers[1] = test_samplers[1].reset()
+
+@thread_wrapped_func
+def train_mp(args, model, train_sampler, valid_samplers=None, rank=0, rel_parts=None, barrier=None):
+    if args.num_proc > 1:
+        th.set_num_threads(4)
```
Comment (zheng-da) [inline apps/kg/train_pytorch.py]: I guess we need to make it more automatic. This setup is for multi-GPU training on p3 instances. It's not optimal for multi-CPU training or other GPU instances.

File: src/graph/sampler.cc (hunk 2 above)
Comment (zheng-da) [reply src/graph/sampler.cc]: agreed.

Comment (zheng-da) [reply src/graph/sampler.cc]: we should remove this change from this PR. We should separate DGL-ke modifications from DGL core modifications. we might need to create a minor version release of DGL core for DGL app packages. To create a minor version release, it's usually cherry-pick commits from the master branch. Having the modification in this PR makes it difficult to create minor release of DGL.

File: apps/kg/dataloader/sampler.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply apps/kg/dataloader/sampler.py]: can we make it larger? something like 32? it doesn't increase too much overhead. The main overhead is more memory to store mini-batch graph structure.

[the remark this answers is not in the document]
Comment (classicsong) [reply apps/kg/dataloader/sampler.py]: OK

File: apps/kg/models/pytorch/tensor_models.py (hunk 8 above)
Comment (classicsong) [reply apps/kg/models/pytorch/tensor_models.py]: relplication of which one? you means the async update?
I think it is OK. More thing will be added here if we support soft relation partition. I am working on it.

File: apps/kg/train_pytorch.py (hunk 9 above)
Comment (classicsong) [reply apps/kg/train_pytorch.py]: i c. Maybe we can provide a arg and give automatic value for multi-CPU and GPU.

File: src/graph/sampler.cc (hunk 2 above)
Comment (classicsong) [reply src/graph/sampler.cc]: This modification is necessary. The sampler semantic has changed slightly.

File: apps/kg/train_pytorch.py [hunk 10]
```diff
@@ -90,52 +68,78 @@ def train(args, model, train_sampler, rank=0, rel_parts=None, valid_samplers=Non
...
     test_samplers[1] = test_samplers[1].reset()
+
+@thread_wrapped_func
+def train_mp(args, model, train_sampler, valid_samplers=None, rank=0, rel_parts=None, barrier=None):
+    if args.num_proc > 1:
+        th.set_num_threads(args.num_thread)
+    train(args, model, train_sampler, valid_samplers, rank, rel_parts, barrier)
+
+@thread_wrapped_func
+def test_mp(args, model, test_samplers, rank=0, mode='Test', queue=None):
```
Comment (zheng-da) [inline apps/kg/train_pytorch.py]: should we set OMP threads here as well?
